### DISCUSSION TITLE
Need a way how to completely disable Copilot per project

### DISCUSSION THREAD

**Discussion by bwajtr** [community/community#73584] [Copilot Conversations]
### Select Topic Area

Product Feedback

### Body

We are using Copilot Business in Intellij Idea. We are working on a number of different projects - contracts of some of them allow us to use Copilot, some of them don't. 

Currently, we have to enable/disable the Copilot plugin depending on which project we work on... Please create a per-project IDE setting allowing us to configure Copilot per-project. Note that "Disable completions" option is not working for us... because we switch the projects so often that sometimes we forget to disable... and also because even with completions disabled you can still trigger Copilot manually ... and that's unwanted behavior...

Or something like `.copilotignore` file - supported by all the plugins - would be even better, so the configuration is done once for all participants on the project. Note that JetBrains AI Assistant is looking for `.noai` file in the root directory of the project and if found, the AI is disabled. It would be nice if the GitHub copilot plugin supported this filename too

**Comment by DJLink** [comment]
a .copilotignore would be great. but I'm thinking, considering that I think we can run shell commands per project, if there's a way to disable a certain plugin by command line into intellij IDE, but that's probably not very reliable, but could be a temporary hack if it works.

**Comment by Buslowicz** [comment]
Sign me up for that! Was just searching for this exact feature!

**Comment by bwajtr** [comment]
In the meantime, JetBrains released their AI Assistant. They decided to add support to disable AI per project. One of the way how to do it is to add `.noai` file to the root directory of the project. Please... **please** support the same file too.
https://www.jetbrains.com/help/idea/disable-ai-assistant.html#restrict-ai-assistant-usage-for-project

I've added it to the body of the description too

**Comment by moezzie** [comment]
This is a must have in my opinion. Some projects you just cannot risk having code leaking out of.

**Comment by twelsh37** [comment]
Is there any fix for this for third party AI Tools?

I have uninstalled the Github Copilot in my project as I wish to use the JetBrains AI assistant. Now I am getting the error below.

![image](https://github.com/community/community/assets/4956770/dc86f8b1-e874-44a4-a6e2-dde5828c2c46)


Mostly Annoying and I'm sure I can just ignore it. But that's just a niggle to address another day.

**Reply by RobLW** [reply]
Just out of interest, can I ask why you are choosing JetBrains AI over Github AI?

**Comment by jakemayfield** [comment]
Very disappointing to see that this has gone unanswered for so long now.

**Reply by twelsh37** [reply]
Sorry, I was lax on This. JetBrains came back to me on their forum with the resolution. It worked perfectly for my issue.

Please check if Copilot is set under File | Settings > Build, Execution, Deployment > Required Plugins and remove it from the list if it is there:

![image](https://github.com/community/community/assets/4956770/72933bd9-eb2b-47d9-8af1-abf6068eb8bf)

**Comment by DJLink** [comment]
I think this might help, configuring content exclusions for github copilot

https://docs.github.com/en/copilot/managing-copilot-business/configuring-content-exclusions-for-github-copilot

**Reply by runningrandall** [reply]
This works great, especially for business/enterprise organizations.

**Comment by twelsh37** [comment]
Hey Rob,
In Answer as to why I chose Jetbrains AI over Github Copilit. Please see below.

I have used copilot for the past year and it has been a good experience. I
am now trying jetbrains AI as I develop in pyCharm, it is only logical that
I asses jetbrains AI.

I might find that I come back to GitHub copilot.

I had already tried Bito, but I was getting better answers from
codellama-34b-instruct model.

For me it is all about testing and comparing.

Regards

Tom

**Comment by VDAB-Wim** [comment]
I think a standard for enabling/disabling AI assistants in a project is a must-have (I'm not just talking about github copilot).
`.noai` is a start, but something like `.aiexclude` is much more interesting (android studio).

While content exclusions work, I have come to understand from the people managing that, that there are limits w.r.t. the amount of exclusions they can configure. Additionally, the exclusions are not tied to the repository itself; i.e. with a copilot license from outside the github organization, they would not apply. Neither would they apply when using another AI assistant...
Especially in organisations that make extensive use of external devs and mixed teams, this is really a must-have. Also for consulting organisations that provide these external devs...

While github copilot (and others) are extremely useful tools, the lack of this feature is a sign of the relative immaturity of these tools.
I guess standardization for this will take time... also considering the prospect of local models, this standardization may be premature as well...
I propose `.localai` as well :smile: .

Kind Regards,

Wim

**Comment by wilson0x4d** [comment]
`.aiignore` following the format and spec of `.gitignore` would be nice, familiar to many.

walking up the folder hierarchy (parent traversal) looking for `.aiignore` would also be nice, because one could be created for entire collections of repositories/folders.

just a thought.

**Reply by wilson0x4d** [reply]
Food for thought if a Product Owner or Project Manager ever stumbles upon this:

Because of the current design for content exclusions we've instead adopted `Continue.dev` and `ollama` to provide coding assistance, and we explicitly disallow the use of Copilot. We are also evaluating tools like `Cline` and adoption of open-source MCP servers for agentic behaviors, but so far Continue+ollama has served as a reasonable alternative to Copilot -- unfortunately with Cline we're still working out some kinks with getting instruct models to behave consistently and not hallucinate.

This because we privately host most of our repositories (Gitea instead of Github) to protect IP, and the entrenchment-based model of Copilot simply didn't work for us (there was no obvious, org-wide method for us to prevent Copilot from ingesting content except to simply not enable it -- because it's inextricably attached to Github auth, a very unfortunate design decision reminiscent of pre-FOSS Microsoft.)

**Comment by stoerr** [comment]
Does anybody know where to make such suggestions? I've never seen any reply by any Github Copilot employee about this kind of request. Not sure whether they actually read that.
